DatadogRUM/Sources/SDKMetrics/SessionEndedMetricController.swift

```diff
@@ -0,0 +1,65 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+import DatadogInternal
+
+/// A controller responsible for managing "RUM Session Ended" metrics.
+internal final class SessionEndedMetricController {
+    /// Dictionary to keep track of pending metrics, keyed by session ID.
+    @ReadWriteLock
+    private var metricsBySessionID: [String: SessionEndedMetric] = [:]
+    /// Array to keep track of pending metrics in their start order.
+    @ReadWriteLock
+    private var metrics: [SessionEndedMetric] = []
```

Comment:
/discussion Mutating a metric property will acquire at least 2 locks, 1 in the controller and one in the metric. The perf penalty will be minimal because we don't expect many concurrent calls. But I want to highlight that the `@ReadWriteLock` wrapper uses a fair pthread lock which is not cheap. 

Moreover, the `metric.track(*_:)` methods are acquiring the property lock multiple times, which could be prevented by calling the [`mutate`](https://github.com/DataDog/dd-sdk-ios/blob/develop/DatadogInternal/Sources/Concurrency/ReadWriteLock.swift#L49) closure.

I think it could be redesigned to use a single lock in the controller and perform all mutations at the controller level, something like the following where `SessionEndedMetric` could be a `struct`:

```swift
internal final class SessionEndedMetricController {
    // Starts a new metric for a given session.
    func startMetric(sessionID: String, precondition: RUMSessionPrecondition?, context: DatadogContext)

    // Tracks the view event that occurred during the session.
    func track(view: RUMViewEvent, for sessionID: String)

    // Tracks the kind of SDK error that occurred during the session.
    func track(sdkErrorKind: String, for sessionID: String)

    // Signals that the session was stopped with `stopSession()` API.
    func trackWasStopped(for sessionID: String)

    // Ends the metric for a given session, sending it to telemetry and removing it from pending metrics.
    func endMetric(sessionID: String)

    // Retrieves the last started metric.
    var latestMetric: SessionEndedMetric? { get }
}
```
What do you think?

Note: the `@ReadWriteLock` is convenient but shouldn't be our only lock. We could provide a lock object as well (similar to NSLock) so we can acquire a lock for operations and not only for property access. We could also provide unfair lock such as `os_unfair_lock` which are cheap.

Reply:
👍 I had this concern on double locking, but ignored it with assuming minimal impact and trying to keep the solution simple. With this feedback I'm happy to improve this 👌.

Reply:
✅ Addressed 👍. I made the `SessionEndedMetric` a `struct` with mutable methods. The controller is now a class which mutates the right `SEM` using single RW lock 👍.